core/src/main/java/org/kohsuke/stapler/bind/BoundObjectTable.java

```diff
@@ -333,54 +311,48 @@ public void generateResponse(StaplerRequest2 req, StaplerResponse2 rsp, Object n
      */
     interface Ref extends Serializable {
         Object get();
+
+        boolean valid();
     }
 
     private static class StrongRef implements Ref {
         private final Object o;
+        private final long expiration;
 
         StrongRef(Object o) {
             this.o = o;
+            expiration = System.currentTimeMillis() + EXPIRATION_TIME;
         }
 
         @Override
         public Object get() {
             return o;
         }
 
-        private Object writeReplace() {
-            if (o instanceof Serializable) {
-                return this;
-            } else {
-                LOGGER.fine(() -> "Refusing to serialize " + o);
-                return new StrongRef(null);
-            }
-        }
-    }
-
-    private static class WeakRef extends WeakReference implements Ref {
-        private WeakRef(Object referent) {
-            super(referent);
+        @Override
+        public boolean valid() {
+            return /* compatibility */ expiration == 0 || System.currentTimeMillis() < expiration;
         }
 
         private Object writeReplace() {
-            Object o = get();
             if (o instanceof Serializable) {
                 return this;
             } else {
                 LOGGER.fine(() -> "Refusing to serialize " + o);
-                return new WeakRef(null);
+                return new StrongRef(null);
             }
         }
     }
 
-    public static final String PREFIX = "/$stapler/bound/";
-    static final String SCRIPT_PREFIX = "/$stapler/bound/script";
-
     /**
-     * True to activate debug logging of session fragments.
+     * How long (in milliseconds) a {@link StrongRef} remains valid for creation if it is not retrieved.
      */
-    @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Legacy switch.")
-    public static boolean DEBUG_LOGGING = Boolean.getBoolean(BoundObjectTable.class.getName() + ".debugLog");
+    private static final long EXPIRATION_TIME = Long.getLong(
+            BoundObjectTable.class.getName() + ".EXPIRATION_TIME",
+            Duration.ofMinutes(15).toMillis());
```

Comment:
Seems long enough for practical purposes. After this point, unused render-on-demand fragments will not work. For example, if you opened a project config form and came back 15m later and selected a parameter type from the dropdown, it will not render the per-parameter config snippet.

Reply:
Are we prepared to revert this based on user feedback? Because I'd be surprised if this didn't happen with some regularity.

Reply:
WDYT about https://github.com/jenkinsci/stapler/pull/665 + https://github.com/jenkinsci/jenkins/pull/10631?

Reply:
If it works, sure. Or we can just increase the timeout to several hours, at which point your session is likely to have expired anyway so this would just be a fallback defense.

Reply:
Good point, cleaning up after 24 hrs or so (this approach) should not affect users while my PR keeps memory use low in the mean time, allowing for the greater timeout 👍